pytest-cov makes test fail by putting a temporary file into cwd

i have tests that check if some specific files have been extracted into cwd.
they fail when using pytest-cov because some .coverage* file gets created there.

is there any way to avoid this?

Thanks for the pointer, I tried it:

Still getting this failing test:

>    AssertionError: Lists differ: ['.coverage.testing-worker-lin... != ['file2']

Hmm, maybe this is not a temporary file of coverage, but a lock/pid file of pytest-cov?

--- a/src/pytest_cov/embed.py
+++ b/src/pytest_cov/embed.py
@@ -41,7 +41,8 @@ def init():
 
     cov_source = os.environ.get('COV_CORE_SOURCE')
     cov_config = os.environ.get('COV_CORE_CONFIG')
-    if cov_config:
+    cov_datafile = os.environ.get('COV_CORE_DATAFILE')
+    if cov_datafile:
         # Import what we need to activate coverage.
         import coverage
 
@@ -50,12 +51,17 @@ def init():
             cov_source = None
         else:
             cov_source = cov_source.split(os.pathsep)
+        if not cov_config:
+            cov_config = True
 
         # Activate coverage for this process.
-        cov = coverage.coverage(source=cov_source,
-                                data_suffix=True,
-                                config_file=cov_config,
-                                auto_data=True)
+        cov = coverage.coverage(
+            source=cov_source,
+            data_suffix=True,
+            config_file=cov_config,
+            auto_data=True,
+            data_file=cov_datafile
+        )
         cov.load()
         cov.start()
         cov._warn_no_data = False

--- a/src/pytest_cov/engine.py
+++ b/src/pytest_cov/engine.py
@@ -34,14 +34,21 @@ class CovController(object):
         if self.cov_source is None:
             os.environ['COV_CORE_SOURCE'] = ''
         else:
-            os.environ['COV_CORE_SOURCE'] = os.pathsep.join(self.cov_source)
-        os.environ['COV_CORE_CONFIG'] = self.cov_config
+            os.environ['COV_CORE_SOURCE'] = os.pathsep.join(
+                os.path.abspath(p) for p in self.cov_source)
+        config_file = os.path.abspath(self.cov_config)
+        if os.path.exists(config_file):
+            os.environ['COV_CORE_CONFIG'] = config_file
+        else:
+            os.environ['COV_CORE_CONFIG'] = ''
+        os.environ['COV_CORE_DATAFILE'] = os.path.abspath('.coverage')
 
     @staticmethod
     def unset_env():
         """Remove coverage info from env."""
         os.environ.pop('COV_CORE_SOURCE', None)
         os.environ.pop('COV_CORE_CONFIG', None)
+        os.environ.pop('COV_CORE_DATAFILE', None)
 
     @staticmethod
     def get_node_desc(platform, version_info):

--- a/tests/test_pytest_cov.py
+++ b/tests/test_pytest_cov.py
@@ -7,11 +7,11 @@ from distutils.version import StrictVersion
 import coverage
 import py
 import pytest
-
 import virtualenv
 from process_tests import TestProcess
 from process_tests import dump_on_error
 from process_tests import wait_for_strings
+
 import pytest_cov.plugin
 
 coverage, StrictVersion  # required for skipif mark on test_cov_min_from_coveragerc
@@ -69,6 +69,30 @@ def test_foo(idx):
         [sys.executable, 'child_script.py', str(idx)],
         stdout=subprocess.PIPE,
         stderr=subprocess.PIPE).communicate()
+
+# there is a issue in coverage.py with multiline statements at
+# end of file: https://bitbucket.org/ned/coveragepy/issue/293
+pass
+'''
+
+SCRIPT_PARENT_CHANGE_CWD = '''
+import subprocess
+import sys
+import os
+
+def pytest_generate_tests(metafunc):
+    for i in range(2):
+        metafunc.addcall(funcargs=dict(idx=i))
+
+def test_foo(idx):
+    os.mkdir("foobar")
+    os.chdir("foobar")
+
+    subprocess.check_call([
+        sys.executable,
+        os.path.join(os.path.dirname(__file__), 'child_script.py'),
+        str(idx)
+    ])
 
 # there is a issue in coverage.py with multiline statements at
 # end of file: https://bitbucket.org/ned/coveragepy/issue/293
@@ -346,6 +370,30 @@ def test_central_subprocess(testdir):
         '*- coverage: platform *, python * -*',
         'child_script* %s *' % CHILD_SCRIPT_RESULT,
         'parent_script* %s *' % PARENT_SCRIPT_RESULT,
+    ])
+    assert result.ret == 0
+
+
+def test_central_subprocess_change_cwd(testdir):
+    scripts = testdir.makepyfile(parent_script=SCRIPT_PARENT_CHANGE_CWD,
+                                 child_script=SCRIPT_CHILD)
+    parent_script = scripts.dirpath().join('parent_script.py')
+    testdir.makefile('', coveragerc="""
+[run]
+branch = true
+parallel = true
+""")
+
+    result = testdir.runpytest('-v', '--tb=short',
+                               '--cov=%s' % scripts.dirpath(),
+                               '--cov-config=coveragerc',
+                               '--cov-report=term-missing',
+                               parent_script)
+
+    result.stdout.fnmatch_lines([
+        '*- coverage: platform *, python * -*',
+        'child_script* %s *' % CHILD_SCRIPT_RESULT,
+        'parent_script* 100% *',
     ])
     assert result.ret == 0
 

